working/0546-patterns/patterns-feature-specification.md

```diff
@@ -144,953 +144,1073 @@ double calculateArea(Shape shape) =>
+capturing the suffix.**
 
-mapTypeArguments ::= '<' typeOrBinder ',' typeOrBinder '>'
+### Extractor pattern
 
-mapBinderEntries ::= mapBinderEntry ( ',' mapBinderEntry )* ','?
-mapBinderEntry ::= expression ':' binder
+```
+extractorPattern ::= extractorName typeArguments? '(' patternFields? ')'
+extractorName    ::= typeIdentifier | qualifiedName
 ```
 
-If it is a compile-time error if any of the entry key expressions are not
-constant expressions. It is a compile-time error if any of the entry key
-expressions evaluate to equivalent values.
+An extractor matches values of a given named type and then extracts values from
+it by calling getters on the value. Extractor patterns let users destructure
+data from arbitrary objects using the getters the object's class already
+exposes.
 
-#### Record binder
+This pattern is particularly useful for writing code in an algebraic datatype
+style. For example:
 
-A record pattern destructures fields from records and objects.
+```dart
+class Rect {
+  final double width, height;
 
-```
-recordBinder        ::= '(' recordFieldBinders ')'
+  Rect(this.width, this.height);
+}
 
-recordFieldBinders  ::= recordFieldBinder ( ',' recordFieldBinder )* ','?
-recordFieldBinder   ::= ( identifier ':' )? binder
-                      | identifier ':'
+display(Object obj) {
+  switch (obj) {
+    case Rect(width: w, height: h): print('Rect $w x $h');
+    default: print(obj);
+  }
+}
 ```
 
-Each field is either a binder which destructures a positional field, or a binder
-prefixed with an identifier and `:` which destructures a named field.
-
-Each named field in the record pattern is matched by calling a corresponding
-getter on the matched object. A record pattern can be applied to an object of
-any type with any set of named field patterns as long as the object being
-matched has getters with those names. (Positional fields, however, can only be
-matched by record values.)
-
-When destructuring named fields, it's common to want to bind the resulting value
-to a variable with the same name. As a convenience, the binder can be omitted on
-a named field. In that case, the field implicitly contains a variable binder
-subpattern with the same name. These are equivalent:
+You can also use an extractor to both match an enum value and destructure
+fields from it:
 
 ```dart
-var (first: first, second: second) = (first: 1, second: 2);
-var (first:, second:) = (first: 1, second: 2);
-```
-
-**TODO: Allow a `...` element in order to ignore some positional fields while
-capturing the suffix.**
+enum Severity  {
+  error(1, "Error"),
+  warning(2, "Warning");
 
-#### Wildcard binder
+  Severity(this.level, this.prefix);
 
-A wildcard binder pattern does nothing.
+  final int level;
+  final String prefix;
+}
 
-```
-wildcardBinder ::= "_"
+log(Severity severity, String message) {
+  switch (severity) {
+    case Severity.error(prefix: p):
+      print('!! $p !! $message'.toUppercase());
```

Comment:
That seems a little weird. The `Severity.error.prefix` is a fixed constant.
Why do you want to match the enum, and still bind the prefix, instead of just:
```dart
  case Severity.error: 
      print('!! ${Severity.error.prefix} !! $message'.toUpperCase());
``` 
or, if you really want a shorter way to refer to it:
```dart
  case Severity.error & e: 
      print('!! ${e.prefix} !! $message'.toUpperCase());
```
(The `& e` is one way to get a short name for anything you have matched, even those patterns which don't allow binding themselves).